### ISSUE TITLE
Negative sampling

### ISSUE THREAD

**Issue by whistle9** [intfloat/SimKGC#21]
I want to ask a very simple but important question for me, that is, as a novice, I can't accurately find the three negative sampling parts in the code. Can you help me point out? I need this answer very much. I would be very excited if you can give me an answer

**Comment by intfloat** [comment]
The in-batch negatives simply use samples from the same batch: https://github.com/intfloat/SimKGC/blob/97cc43e488f19ca5b0f6fbf60ffefd2ee56c0693/models.py#L91

The pre-batch negatives use cached negatives from previous batches:
https://github.com/intfloat/SimKGC/blob/97cc43e488f19ca5b0f6fbf60ffefd2ee56c0693/models.py#L117-L133

The self-negatives use the head entities of each example:
https://github.com/intfloat/SimKGC/blob/97cc43e488f19ca5b0f6fbf60ffefd2ee56c0693/models.py#L104-L109

**Comment by whistle9** [comment]
Thank you for your detailed answer. Is all the parameters used in the self-negative sampling method fixed? If I want to use this method in other frameworks, do the parameters used here also have to be？

**Comment by intfloat** [comment]
Not sure what you are referring to... Self-negative sampling does not add any new parameters.

**Comment by whistle9** [comment]
Sorry, I didn't make it clear. I didn't communicate with you in a timely manner due to some issues.
I mean the parameters in the self negative sampling of lines 104-109 of the models.py file, such as head_ vector、self_ neg_ logits... and logits.Are these all necessary parameters?
<img width="678" alt="image" src="https://user-images.githubusercontent.com/40874661/227899719-ab7568a5-5590-490e-a4cc-5cc09d30d187.png">
In addition, I found that the negative sampling part belongs to the compute_ logits function has only been called twice by trainer.py (lines 114/154), respectively eval_epoch function, train_ epoch function, I don't quite understand this part.Could you provide me with an answer? Thank you and look forward to your reply.".
<img width="859" alt="image" src="https://user-images.githubusercontent.com/40874661/227902701-879a9046-e792-4301-86cb-64387526bbe5.png">
<img width="902" alt="image" src="https://user-images.githubusercontent.com/40874661/227902834-36fdba0c-50ec-42b6-95a7-8b6fa5914a28.png">

**Comment by qiaoxiaoqiao1** [comment]
> Not sure what you are referring to... Self-negative sampling does not add any new parameters.

May I know why you set -1e4 for masked_fill?

**Comment by intfloat** [comment]
> > Not sure what you are referring to... Self-negative sampling does not add any new parameters.
> 
> May I know why you set -1e4 for masked_fill?

Any number that is small enough should be able to mask the probability to 0.
